### PR TITLE
Switch to yarn

### LINKED ISSUE #393: Switch from npm to yarn
https://trello.com/c/lAhELsEw/259-switch-from-npm-to-yarn

`yarn` is, apparently the new hotness. Would be nice to switch over, particularly next time `npm` or `npm-shrinkwrap` causes any pain.

### PR DESCRIPTION
**Pull request by outoftime**
Switches to using [Yarn](https://yarnpkg.com/) for package management. Yarn is the cool new NPM.

Changes:

* Swap out `npm-shrinkwrap.json` for `yarn.lock`
* Updates Travis configuration to install and use Yarn
* Updates the Webpack config to deal with different module layout and inadvertently updated packages
  * Specifically, switches to whitelisting stylelint rules at the module loader level. This means that we are more resilient to new rules causing the build to break, and also has the advantage of substantially reducing bundle size
* Updates the README installation instructions

Fixes #393

### CODE REVIEW

**Comment by outoftime** [comment]
@alexpelan @jwang1919 WDYT?

**Comment by jwang1919** [comment]
@outoftime I've never used yarn before. What's your experience so far with it?

**Comment by outoftime** [comment]
@jwang1919 word, it’s really new! but we switch over at Genius a couple of weeks ago and I’m really happy with it. The main selling points are:

* The layout of the `node_modules` directory is deterministic based on the contents of `package.json`
* It’s *much* faster (like, an order of magnitude for a cold install)
* It caches packages (so you can switch branches and run an install without an internet connection)
* All operations update both `package.json` and `yarn.lock` (the latter is the Yarn equivalent of `npm-shrinkwrap.json`)

Big blog post [here](https://code.facebook.com/posts/1840075619545360), and the [home page](https://yarnpkg.com/) has some good TL;DR on why it’s good

**Comment by jwang1919** [comment]
Sounds good. After doing a trial run, I'm getting an error regarding fsevents again...

Stacktrace:
```
error Error running install script for optional dependency: "C:\\Users\\jwang\\Documents\\CodeProjects\\popcode\\node_modules\\fsevents: Command failed.\nExit code: 1\nCommand: C:\\WINDOWS\\system32\\cmd.exe\nArguments: /d /s /c node-pre-gyp install --fallback-to-build\nDirectory: C:\\Users\\jwang\\Documents\\CodeProjects\\popcode\\node_modules\\fsevents\nOutput:\nnode-pre-gyp info it worked if it ends with ok\nnode-pre-gyp info using node-pre-gyp@0.6.29\nnode-pre-gyp info using node@6.8.1 | win32 | x64\nnode-pre-gyp info check checked for \"C:\\Users\\jwang\\Documents\\CodeProjects\\popcode\\node_modules\\fsevents\\lib\\binding\\Release\\node-v48-win32-x64\\fse.node\" (not found)\nnode-pre-gyp http GET https://fsevents-binaries.s3-us-west-2.amazonaws.com/v1.0.15/fse-v1.0.15-node-v48-win32-x64.tar.gz\nnode-pre-gyp http 404 https://fsevents-binaries.s3-us-west-2.amazonaws.com/v1.0.15/fse-v1.0.15-node-v48-win32-x64.tar.gz\nnode-pre-gyp ERR! Tried to download: https://fsevents-binaries.s3-us-west-2.amazonaws.com/v1.0.15/fse-v1.0.15-node-v48-win32-x64.tar.gz \nnode-pre-gyp ERR! Pre-built binaries not found for fsevents@1.0.15 and node@6.8.1 (node-v48 ABI) (falling back to source compile with node-gyp) \nnode-pre-gyp http Pre-built binary not available for your system, looked for https://fsevents-binaries.s3-us-west-2.amazonaws.com/v1.0.15/fse-v1.0.15-node-v48-win32-x64.tar.gz \n\r\nC:\\Users\\jwang\\Documents\\CodeProjects\\popcode\\node_modules\\fsevents>if not defined npm_config_node_gyp (node \"C:\\Users\\jwang\\AppData\\Roaming\\npm\\node_modules\\yarn\\bin\\node-gyp-bin\\\\..\\..\\node_modules\\node-gyp\\bin\\node-gyp.js\" clean )  else (node  clean ) \r\ngyp info it worked if it ends with ok\ngyp info using node-gyp@3.4.0\ngyp info using node@6.8.1 | win32 | x64\ngyp info ok \n\r\nC:\\Users\\jwang\\Documents\\CodeProjects\\popcode\\node_modules\\fsevents>if not defined npm_config_node_gyp (node \"C:\\Users\\jwang\\AppData\\Roaming\\npm\\node_modules\\yarn\\bin\\node-gyp-bin\\\\..\\..\\node_modules\\node-gyp\\bin\\node-gyp.js\" configure --fallback-to-build --module=C:\\Users\\jwang\\Documents\\CodeProjects\\popcode\\node_modules\\fsevents\\lib\\binding\\Release\\node-v48-win32-x64\\fse.node --module_name=fse --module_path=C:\\Users\\jwang\\Documents\\CodeProjects\\popcode\\node_modules\\fsevents\\lib\\binding\\Release\\node-v48-win32-x64 )  else (node  configure --fallback-to-build --module=C:\\Users\\jwang\\Documents\\CodeProjects\\popcode\\node_modules\\fsevents\\lib\\binding\\Release\\node-v48-win32-x64\\fse.node --module_name=fse --module_path=C:\\Users\\jwang\\Documents\\CodeProjects\\popcode\\node_modules\\fsevents\\lib\\binding\\Release\\node-v48-win32-x64 ) \r\ngyp info it worked if it ends with ok\ngyp info using node-gyp@3.4.0\ngyp info using node@6.8.1 | win32 | x64\ngyp ERR! configure error \ngyp ERR! stack Error: Can't find Python executable \"python\", you can set the PYTHON env variable.\ngyp ERR! stack     at failNoPython (C:\\Users\\jwang\\AppData\\Roaming\\npm\\node_modules\\yarn\\node_modules\\node-gyp\\lib\\configure.js:449:14)\ngyp ERR! stack     at C:\\Users\\jwang\\AppData\\Roaming\\npm\\node_modules\\yarn\\node_modules\\node-gyp\\lib\\configure.js:404:11\ngyp ERR! stack     at C:\\Users\\jwang\\AppData\\Roaming\\npm\\node_modules\\yarn\\node_modules\\graceful-fs\\polyfills.js:276:29\ngyp ERR! stack     at FSReqWrap.oncomplete (fs.js:123:15)\ngyp ERR! System Windows_NT 10.0.14393\ngyp ERR! command \"C:\\\\Program Files\\\\nodejs\\\\node.exe\" \"C:\\\\Users\\\\jwang\\\\AppData\\\\Roaming\\\\npm\\\\node_modules\\\\yarn\\\\node_modules\\\\node-gyp\\\\bin\\\\node-gyp.js\" \"configure\" \"--fallback-to-build\" \"--module=C:\\\\Users\\\\jwang\\\\Documents\\\\CodeProjects\\\\popcode\\\\node_modules\\\\fsevents\\\\lib\\\\binding\\\\Release\\\\node-v48-win32-x64\\\\fse.node\" \"--module_name=fse\" \"--module_path=C:\\\\Users\\\\jwang\\\\Documents\\\\CodeProjects\\\\popcode\\\\node_modules\\\\fsevents\\\\lib\\\\binding\\\\Release\\\\node-v48-win32-x64\"\ngyp ERR! cwd C:\\Users\\jwang\\Documents\\CodeProjects\\popcode\\node_modules\\fsevents\ngyp ERR! node -v v6.8.1\ngyp ERR! node-gyp -v v3.4.0\ngyp ERR! not ok \nnode-pre-gyp ERR! build error \nnode-pre-gyp ERR! stack Error: Failed to execute 'node-gyp.cmd configure --fallback-to-build --module=C:\\Users\\jwang\\Documents\\CodeProjects\\popcode\\node_modules\\fsevents\\lib\\binding\\Release\\node-v48-win32-x64\\fse.node --module_name=fse --module_path=C:\\Users\\jwang\\Documents\\CodeProjects\\popcode\\node_modules\\fsevents\\lib\\binding\\Release\\node-v48-win32-x64' (1)\nnode-pre-gyp ERR! stack     at ChildProcess.<anonymous> (C:\\Users\\jwang\\Documents\\CodeProjects\\popcode\\node_modules\\fsevents\\node_modules\\node-pre-gyp\\lib\\util\\compile.js:83:29)\nnode-pre-gyp ERR! stack     at emitTwo (events.js:106:13)\nnode-pre-gyp ERR! stack     at ChildProcess.emit (events.js:191:7)\nnode-pre-gyp ERR! stack     at maybeClose (internal/child_process.js:877:16)\nnode-pre-gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:226:5)\nnode-pre-gyp ERR! System Windows_NT 10.0.14393\nnode-pre-gyp ERR! command \"C:\\\\Program Files\\\\nodejs\\\\node.exe\" \"C:\\\\Users\\\\jwang\\\\Documents\\\\CodeProjects\\\\popcode\\\\node_modules\\\\fsevents\\\\node_modules\\\\node-pre-gyp\\\\bin\\\\node-pre-gyp\" \"install\" \"--fallback-to-build\"\nnode-pre-gyp ERR! cwd C:\\Users\\jwang\\Documents\\CodeProjects\\popcode\\node_modules\\fsevents\nnode-pre-gyp ERR! node -v v6.8.1\nnode-pre-gyp ERR! node-pre-gyp -v v0.6.29\nnode-pre-gyp ERR! not o$ ./node_modules/.bin/bower install
'.' is not recognized as an internal or external command,
operable program or batch file.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

**Comment by jwang1919** [comment]
Arggh still couldn't get it to work. Tried removing `yarn.lock`, adding `.yarnrc` file with `ignore-optional false`.

Also found that the issue might not have been fixed: https://github.com/yarnpkg/yarn/issues/1059

**Comment by outoftime** [comment]
@jwang1919 https://github.com/yarnpkg/yarn/issues/1059 suggests that maybe `ignore-scripts=true` might fix it?

**Comment by jwang1919** [comment]
@outoftime Tried adding `ignore-scripts true` (yarn doesn't recognize `ignore-scripts=true`).
What it's doing is hiding the error but not resolving the actual issue. The same error is happening when I run `gulp dev`, the server starts throwing errors that it can't find the `fs` library.

*sigh* Windows problems =(

**Comment by outoftime** [comment]
Damn @jwang1919 that _really_ sucks! Yarn is so much better!

OK well I will put this on hold and keep an eye on https://github.com/yarnpkg/yarn/issues/1059

**Comment by outoftime** [comment]
@jwang1919 I don’t suppose you have any interest in developing in an Ubuntu VM : )

File: yarn.lock [hunk 1]
```diff
@@ -0,0 +1,7589 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (jwang1919) [inline yarn.lock]: Regenerate this file. After updating to yarn v.0.18.0 this file changed.

File: package.json [hunk 2]
```diff
@@ -136,13 +136,14 @@
     "karma-webpack": "^1.8.0",
```
Comment (jwang1919) [inline package.json]: Review is not for this line of code but for lines 87-93. I had to change to the following this get it to work:
```
  "scripts": {
    "postinstall": "\"\"./node_modules/.bin/bower\"\" install",
    "pretest": "npm run lint",
    "test": "\"\"./node_modules/.bin/karma\"\" start --single-run --no-auto-watch",
    "autotest": "\"\"./node_modules/.bin/karma\"\" start --no-single-run --auto-watch",
    "lint": "\"\"./node_modules/.bin/eslint\"\" --ext .js,.jsx --plugin react src"
  },
```

Comment (jwang1919) [inline package.json]: Nevermind, your recent changes fixed this issue.

File: yarn.lock (hunk 1 above)
Comment (jwang1919) [inline yarn.lock]: disregard.

**Comment by jwang1919** [comment]
Okay after much waiting I can confirm that upgrading yarn to v0.18.0 will indeed pass `yarn install`!
Note to early Windows adopters, make sure you upgrade to v0.18.0. I had to go directly to the [releases tab on github yarn](https://github.com/yarnpkg/yarn/releases/tag/v0.18.0) to get it. 
`npm -global update yarn` will not update to the right version, neither will downloading the Windows installer from the official yarn site.

**Pull request merged**